diff --git a/src/models/gemma3.cpp b/src/models/gemma3.cpp
index 63a2b380e..b982bd9c6 100644
--- a/src/models/gemma3.cpp
+++ b/src/models/gemma3.cpp
@@ -1,6 +1,8 @@
 #include "models.h"

 void llama_model_gemma3::load_arch_hparams(llama_model_loader & ml) {
+    const bool found_attention_scale =
+        ml.get_key(LLM_KV_ATTENTION_SCALE, hparams.f_attention_scale, false);
     const bool found_swa = ml.get_key(LLM_KV_ATTENTION_SLIDING_WINDOW, hparams.n_swa, false);
     if (found_swa && hparams.n_swa > 0) {
         hparams.swa_type = LLAMA_SWA_TYPE_STANDARD;
@@ -28,9 +30,11 @@ void llama_model_gemma3::load_arch_hparams(llama_model_loader & ml) {
     }

     // ref: https://github.com/google/gemma_pytorch/blob/014acb7ac4563a5f77c76d7ff98f31b568c16508/gemma/config.py#L289
-    hparams.f_attention_scale = type == LLM_TYPE_27B
-        ? 1.0f / std::sqrt(float(hparams.n_embd / hparams.n_head(0)))
-        : 1.0f / std::sqrt(float(hparams.n_embd_head_k()));
+    if (!found_attention_scale || hparams.f_attention_scale <= 0.0f) {
+        hparams.f_attention_scale = type == LLM_TYPE_27B
+            ? 1.0f / std::sqrt(float(hparams.n_embd / hparams.n_head(0)))
+            : 1.0f / std::sqrt(float(hparams.n_embd_head_k()));
+    }
 }

 void llama_model_gemma3::load_arch_tensors(llama_model_loader &) {
